src/io_family.f90

```diff
@@ -0,0 +1,586 @@
+      type(adios2_io) :: io                                   ! adios2 only
+#endif
+   contains
+      procedure :: init => d2d_io_family_init                     ! ADIOS2 writer if possible, MPI otherwise
+      procedure :: mpi_init => d2d_io_family_mpi_init             ! Force MPI writer
+      procedure :: adios2_init => d2d_io_family_adios2_init       ! Force ADIOS2 writer
+      procedure :: fin => d2d_io_family_fin                       ! Clear the writer
+      procedure :: register_var3d => d2d_io_family_register_var3d ! Register a 3D variable
+      procedure :: register_var2d => d2d_io_family_register_var2d ! Register a 2D variable
+      procedure :: log => d2d_io_family_log                       ! Print some information about the object
+   end type d2d_io_family
+
+   private
+
+   public :: decomp_2d_io_family_init, &
+             decomp_2d_io_family_fin
+#ifdef ADIOS2
+   public :: decomp_2d_io_family_get_default_adios, &
+             decomp_2d_io_family_set_default_adios
+#endif
+
+contains
+
+   !
+   ! Initialize a new family of readers / writers (default type)
+   !
+   subroutine d2d_io_family_init(family, label)
+
+      implicit none
+
+      class(d2d_io_family), intent(inout) :: family
+      character(len=*), intent(in) :: label
+
+#ifdef ADIOS2
+      call family%adios2_init(label)
+#else
+      call family%mpi_init(label)
+#endif
+
+   end subroutine d2d_io_family_init
+
+   !
+   ! Initialize a new family of MPI readers / writers
+   !
+   subroutine d2d_io_family_mpi_init(family, label)
+
+      implicit none
+
+      class(d2d_io_family), intent(inout) :: family
+      character(len=*), intent(in) :: label
+
+#ifdef PROFILER
+      if (decomp_profiler_io) call decomp_profiler_start("d2d_io_family_mpi_init")
+#endif
+
+      ! Safety check
+      if (family%type /= decomp_2d_io_none) then
+         call decomp_2d_abort(__FILE__, __LINE__, -1, "Family was not cleared "//label)
+      end if
+
+      family%type = decomp_2d_io_mpi
+      family%label = label
+
+      if (decomp_debug >= D2D_DEBUG_LEVEL_INFO) call family%log()
+
+#ifdef PROFILER
+      if (decomp_profiler_io) call decomp_profiler_end("d2d_io_family_mpi_init")
+#endif
+
+   end subroutine d2d_io_family_mpi_init
+
+   !
+   ! Initialize a new family of ADIOS2 readers / writers
+   !
+   subroutine d2d_io_family_adios2_init(family, label)
+
+      implicit none
+
+      class(d2d_io_family), intent(inout) :: family
+      character(len=*), intent(in) :: label
+
+#ifdef ADIOS2
+      integer :: ierror
+#endif
+
+#ifdef PROFILER
+      if (decomp_profiler_io) call decomp_profiler_start("d2d_io_family_adios2_init")
+#endif
+
+      ! Safety check
+      if (family%type /= decomp_2d_io_none) then
+         call decomp_2d_abort(__FILE__, __LINE__, -1, "Family was not cleared "//label)
+      end if
+
+#ifndef ADIOS2
+      call decomp_2d_abort(__FILE__, __LINE__, -1, "ADIOS2 is not available")
+#else
```

Comment:
I'd put this at the top of the subroutine, no point doing anything further if ADIOS2 isn't available when this is called.

Reply:
https://github.com/2decomp-fft/2decomp-fft/pull/269/commits/ce1aeabac86595c5204cf2a89b3884fecd086f26